---
title: "CD56dim subclustering"
output: html_document
---

````{r setup, include=FALSE, warning=FALSE, message=FALSE, error=FALSE}
library(scater)
library(scran)
library(pheatmap)
library(reshape2)
```


## Import clustered sce object
```{r import object}
sce <- readRDS('sce_clustered.rds')
```

## subclustering CD56dim NK cells
```{r recluster}
sce.dim <- sce[,sce$label %in% c('1')]

#variable genes
dec.dim<- modelGeneVar(sce.dim)
hvg.sce.var <- getTopHVGs(dec.dim, n=2000)
str(hvg.sce.var)

#PCA
set.seed(100) # See below.
sce.dim <- runPCA(sce.dim, subset_row=hvg.sce.var, name="PCA")
percent.var <- attr(reducedDim(sce.dim), "percentVar")
plot(percent.var, xlab="PC", ylab="Variance explained (%)")
reducedDim(sce.dim, "PCA_30") <- reducedDim(sce.dim, "PCA")[,1:30]

#clustering 30PCs otherwise NKG2C population is not clustered
g.dim <- buildSNNGraph(sce.dim, k=10, use.dimred="PCA_30")
clust.dim <- igraph::cluster_infomap(g.dim)$membership
sce.dim$sublabel<- factor(clust.dim)

#UMAP
set.seed(110011001)
sce.dim <- runUMAP(sce.dim, dimred="PCA_30", n_neighbors=20, min_dist=0.1, name="UMAP")
plotReducedDim(sce.dim,dimred="UMAP", colour_by ="sublabel", point_size=2, text_by = 'sublabel')+ theme(text = element_text(size = 13), axis.text=element_text(size=13)) +
scale_color_manual(values = c('1' = "#A0CBE8", '2' = "#D37295", '3' = "#8CD17D", '4'='#F1CE63', '5'='#79706E', '6'='#499894'),aesthetics = "fill")

plotReducedDim(sce.dim,dimred="UMAP", colour_by ="KLRB1", point_size=2, text_by = 'sublabel')+ theme(text = element_text(size = 15), axis.text=element_text(size=15))


ggsave("KLRB1.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 5,
  dpi = 300,limitsize = TRUE,bg = NULL,)

```

##save object
```{r}
saveRDS(sce.dim, 'sce_dim.rds')
```

## Differential gene expression

```{r find differentially expressed genes}
markers<- findMarkers(sce.dim, test="wilcox", groups=sce.dim$sublabel)

cluster1<-as.data.frame(markers[[1]])
cluster2<-as.data.frame(markers[[2]])
cluster3<-as.data.frame(markers[[3]])
cluster4<-as.data.frame(markers[[4]])
cluster5<-as.data.frame(markers[[5]])
cluster6<-as.data.frame(markers[[6]])


#complete heatmap of top 10 most differentially expressed genes, Z-score
top10=list()
for (i in 1:length(markers)){
  top10[[i]] <- rownames(markers[[i]])[markers[[i]]$Top<=10]
}
p <- plotDots(sce.dim, features=unique(unlist(top10)), group='sublabel')
p <- p$data[,c(1,2,4)]
p2 <- dcast(p,Feature ~Group,  value.var='Average')
row.names(p2) <- p2$Feature
p2 <- p2[,-1]
pheatmap(p2, scale="row", cluster_cols=FALSE, cluster_rows = TRUE, cellwidth=7, cellheight = 7, fontsize_row=7, fontsize_col=7)
```

## KIR analysis

```{r KIR expression}
#add sublabel sce.dim to sce
sce.dim$sublabel <- paste('CD56dimGZMKmin',sep="_",sce.dim$sublabel)
sce$sublabel <- sce$subset
sce[,sce$label=='1']$sublabel <- sce.dim$sublabel

#find KIR genes
kirs <- grep("^KIR", rownames(sce))
rownames(sce)[kirs]

kirs <- c("KIR2DL1", "KIR3DL2", "KIR2DL3", "KIR3DL1", "KIR3DL3") #KIR2DL4 excluded = activating KIR

#detection_limit=0, based on the logcounts matrix
sce <- addPerCellQC(sce, subsets=list(KIRS=kirs), exprs_values='logcounts', detection_limit=0)
plotReducedDim(sce, dimred="UMAP", colour_by="subsets_KIRS_detected", text_by="label")
plotColData(sce, y="subsets_KIRS_detected", x='label', colour_by = 'label')


#data for bargraph with # of KIRs
kirs_detected <- plotColData(sce, y="subsets_KIRS_detected", x='sublabel', colour_by = 'sublabel')
kirs_detected <- kirs_detected$data
colnames(kirs_detected) <- c('nrKIRS', "label", "label")
kirs_detected <- kirs_detected[,c(1,2)]
kirs_detected <- table(kirs_detected)
kirs_detected <- prop.table(kirs_detected, margin=2)
barplot(kirs_detected, main="nKIRS detected",xlab="label")

#data for bargraph with inhibitory receptors
p<-plotDots(sce, features=c("KIR2DL1", "KIR3DL2", "KIR2DL3", "KIR3DL1", "KIR3DL3", "KLRC1"), group='sublabel')+geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0.5)
p$data

#data for violin plot relation #KIR and NKG2A
sce$subsets_KIRS_detected <- as.factor(sce$subsets_KIRS_detected)
plotExpression(sce[, sce$label%in%c(1)], features="KLRC1", x="subsets_KIRS_detected", colour_by="subset")+ theme(text = element_text(size = 15),axis.text=element_text(size=13))
ggsave("nkg2a_KIR.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 4, units = c("in", "cm", "mm", "px"),dpi =300,limitsize = TRUE,bg = NULL,)

#check how many cells express NKG2A in relation to # of KIRS detected
p<-plotDots(sce[,sce$label%in%c('1')], features=c("KLRC1"), group='subsets_KIRS_detected')
p$data

```

