---
title: "quality control"
output: html_document
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE, error=FALSE}
library(scater)
library(scran)
library(DropletUtils)
library(Seurat)
library(EnsDb.Hsapiens.v86)
```

## load dataset

```{r loaddataset}
#read as HDF5
sce_unfiltered <- read10xCounts('P:/kjc/Immlab/janineM/10xGenomics/Janine_GEX_count_7k/outs/filtered_feature_bc_matrix.h5', type='HDF5')

sce_unfiltered$Donor <- 'BMKJC'

rowData(sce_unfiltered)$Chr <- mapIds(EnsDb.Hsapiens.v86, keys=rownames(sce_unfiltered),column="SEQNAME", keytype="GENEID")

rownames(sce_unfiltered) <- uniquifyFeatureNames(rowData(sce_unfiltered)$ID,
    names = rowData(sce_unfiltered)$Symbol)

#split ADT
sce_unfiltered <- splitAltExps(sce_unfiltered, rowData(sce_unfiltered)$Type)

#coerce sparse matrix into dense matrix for adt counts
counts(altExp(sce_unfiltered)) <- as.matrix(counts(altExp(sce_unfiltered)))
counts(altExp(sce_unfiltered))[,1:10] # sneak peek
```

## add metadata
```{r}
# add mitodata
mito <- grep("^MT", rowData(sce_unfiltered)$Symbol)
sce_unfiltered <- addPerCellQC(sce_unfiltered, subsets=list(Mito=mito))
colData(sce_unfiltered)[, 1:5]

sce_unfiltered <- addPerFeatureQC(sce_unfiltered)
rowData(sce_unfiltered)

# add ribodata
ribo <- grep("^RP[SL]", rowData(sce_unfiltered)$Symbol)
sce_unfiltered <- addPerCellQC(sce_unfiltered, subsets=list(Ribo=ribo))
colData(sce_unfiltered)

# define discard cells
df <- colData(sce_unfiltered)
qc.nexprs <- df$detected < 1e3 | df$detected >8300
qc.mito <- df$subsets_Mito_percent > 12.5
discard <- qc.nexprs | qc.mito
sce_unfiltered$discard <- discard


# generate plots to highlight discarded cells
gridExtra::grid.arrange(
    plotColData(sce_unfiltered, y="sum", colour_by="discard")+ scale_y_log10() + ggtitle("Total RNA")+ylab("Total RNA"),
    plotColData(sce_unfiltered, y="detected", colour_by="discard") + scale_y_log10() + ggtitle("Total genes") +ylab("Total genes"),
    plotColData(sce_unfiltered,y="subsets_Mito_percent", colour_by="discard") + ggtitle("% Mito")+ylab("% Mito"),
    plotColData(sce_unfiltered,y="subsets_Ribo_percent", colour_by="discard") + ggtitle("% Ribo")+ylab("% Mito"),
    ncol=4
)

plotColData(sce_unfiltered, x="sum", y="detected", colour_by="discard") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+scale_y_continuous(breaks = c(0, 1000, 2500, 5000, 7500, 8300, 10000), limits = c(0, 10000))+geom_hline(yintercept = c(8300,1000), linetype='dashed')+ylab("total RNA")+xlab("Genes")


plotColData(sce_unfiltered, x="sum", y="subsets_Mito_percent", colour_by="discard") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+geom_hline(yintercept = 12.5, linetype = 'dashed')+scale_y_continuous(breaks = c(0, 12.5, 25, 50, 75, 100))+xlab("total RNA")+ylab("% mito")

plotColData(sce_unfiltered, x="sum", y="subsets_Ribo_percent", colour_by="discard") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+xlab("total RNA")+ylab("% ribo")


#remove discarded cells
sce <- sce_unfiltered[,!sce_unfiltered$discard]

```

## normalization of ADT data (CLR by seurat)
```{r}
colnames(sce) <- sce$Barcode
counts <- counts(altExp(sce))
adt.seurat <- CreateSeuratObject(counts = counts, project = "BMKJC", min.cells = 0)            
adt.seurat <- NormalizeData(object = adt.seurat, assay = "RNA", normalization.method = "CLR")

FeatureScatter(object = adt.seurat, feature1 = "CD56-TotalA", feature2 = "CD117-TotalA", pt.size = 1)

#revert back to sce
altExp(sce) <- as.SingleCellExperiment(adt.seurat)
```

## normalization of gene expression data for cell-specific biases
```{r normalization by deconvolution}
#composition biases are introduced by strong differential expression between cell types. Use of the deconvolution size factors adjusts for these biases to improve normalization accuracy for downstream applications
set.seed(1000)
clusters <- quickCluster(sce)
table(clusters)

#add deconv factor to sce
sce <- computeSumFactors(sce, clusters=clusters)
deconv.factor <- calculateSumFactors(sce, clusters=clusters)

#We then compute log-normalized expression values for downstream use.
sce <- logNormCounts(sce, size_factors=deconv.factor)
assayNames(sce)
```

We now continue with clustering and visualization to determine whether the quality control was successfull.

## Select Variable features
```{r}
dec <- modelGeneVar(sce)

# Visualizing the fit:
fit.sce <- metadata(dec)
plot(fit.sce$mean, fit.sce$var, xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fit.sce$trend(x), col="dodgerblue", add=TRUE, lwd=2)

dec[order(dec$bio, decreasing=TRUE),] 

```

## select highly variable genes
```{r}
# Taking the top 2000 genes or top 10%  here:
hvg.sce.var <- getTopHVGs(dec, n=2000)
str(hvg.sce.var)
top10 <- hvg.sce.var[1:10]

plot(fit.sce$mean, fit.sce$var, xlab="Mean of log-expression",
    ylab="Variance of log-expression")
curve(fit.sce$trend(x), col="dodgerblue", add=TRUE, lwd=2)
text(fit.sce$mean[top10], fit.sce$var[top10], labels = names(fit.sce$var[top10]), adj=c(0,1), cex=0.8)

```

##PCA
```{r}
set.seed(100) # See below.
sce <- runPCA(sce, subset_row=hvg.sce.var, name="PCA")
reducedDimNames(sce)
dim(reducedDim(sce, "PCA"))
```

## choose numbers of PCs for subsequent analysis
```{r}
percent.var <- attr(reducedDim(sce, 'PCA'), "percentVar")
plot(percent.var, xlab="PC", ylab="Variance explained (%)")

#apply number of PCS
reducedDim(sce, "PCA_20") <- reducedDim(sce, "PCA")[,1:20]
ncol(reducedDim(sce, "PCA_20"))

#Visualize PCA
plotReducedDim(sce, dimred="PCA", colour_by="SELL")
plotReducedDim(sce, dimred="PCA", colour_by="CD34")
plotReducedDim(sce, dimred="PCA", ncomponents=4)
```

## Clustering
```{r}
#the higher the k the less clusters
g.major <- buildSNNGraph(sce, k=30, use.dimred = 'PCA_20')
clust.major <- igraph::cluster_infomap(g.major)$membership
table(clust.major)
colLabels(sce) <- factor(clust.major)

g.minor <- buildSNNGraph(sce, k=10, use.dimred = 'PCA_20')
clust.minor <- igraph::cluster_infomap(g.minor)$membership
sce$minor <- factor(clust.minor)

plotColData(sce, y="sum", x='label', colour_by="label")+ scale_y_log10() + ggtitle("Total RNA")+ylab("Total RNA")
plotColData(sce, y="sum", x='minor', colour_by="label")+ scale_y_log10() + ggtitle("Total RNA")+ylab("Total RNA")

plotColData(sce, x="sum", y="detected", colour_by="label") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+scale_y_continuous(breaks = c(0, 1000, 2500, 5000, 7500, 8300, 10000), limits = c(0, 10000))+geom_hline(yintercept = c(8300,1000), linetype='dashed')+ylab("total RNA")+xlab("Genes")

plotColData(sce, x="sum", y="subsets_Mito_percent", colour_by="label") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+geom_hline(yintercept = 12.5, linetype = 'dashed')+scale_y_continuous(breaks = c(0, 12.5, 25, 50, 75, 100))+xlab("total RNA")+ylab("% mito")

plotColData(sce, x="sum", y="subsets_Ribo_percent", colour_by="label") +theme(panel.border = element_rect(color = "grey"))+theme_bw()+xlab("total RNA")+ylab("% ribo")

```

## run umap
```{r}
set.seed(110011001)
sce <- runUMAP(sce, dimred="PCA_20", n_neighbors=30, min_dist=0.6)

plotReducedDim(sce, "UMAP", text_by='label', colour_by="label")
plotReducedDim(sce, "UMAP", text_by='minor', colour_by="detected")

```

## remove cells that clustered based on high mRNA content. 

```{r}
sce_filtered <- sce[,sce$minor != '8']

saveRDS(sce_filtered, 'sce_filtered.rds')

```

The quality control is now finished.
