---
title: "pseudotime"
output: html_document
---

````{r setup, include=FALSE, warning=FALSE, message=FALSE, error=FALSE}
library(scater)
library(scran)
library(slingshot)
library(pheatmap)
library(tradeSeq)
library(Seurat)
library(SeuratWrappers)
library(velociraptor)
```

## import required objects
```{r}
sce <- readRDS('sce_clustered.rds')
sce.dim <- readRDS('sce_dim.rds')

# add sublabels to sce object
sce.dim$sublabel <- paste('CD56dimGZMKmin',sep="_",sce.dim$sublabel)
sce$sublabel <- sce$subset
sce[,sce$label=='1']$sublabel <- sce.dim$sublabel

#exclude CD56- cells
sce <- sce[, sce$subset != '8. CD56-' & sce$subset != '7. NKprolif']
```

## rerun PCA and UMAP
```{r}
#variable genes
dec<- modelGeneVar(sce)
hvg.sce.var <- getTopHVGs(dec, n=2000)
str(hvg.sce.var)

#PCA
set.seed(100) # See below.
sce <- runPCA(sce, subset_row=hvg.sce.var, name="PCA")
percent.var <- attr(reducedDim(sce), "percentVar")
plot(percent.var, xlab="PC", ylab="Variance explained (%)")

#UMAP
set.seed(110011001)
sce <- runUMAP(sce, dimred="PCA", n_neighbors=30, min_dist=0.6, name="UMAP")
sce <- runUMAP(sce, dimred="PCA", n_neighbors=30, min_dist=1.0, name="UMAP_1")
sce <- runUMAP(sce, dimred="PCA", n_neighbors=30, min_dist=0.4, name="UMAP_04")
sce <- runUMAP(sce, dimred="PCA", n_neighbors=30, min_dist=0.2, name="UMAP_02")
sce <- runUMAP(sce, dimred="PCA", n_neighbors=30, min_dist=0.1, name="UMAP_01")

plotReducedDim(sce, colour_by="subset", dimred='UMAP')+ theme(text = element_text(size = 15), axis.text=element_text(size=15))
ggsave("umap.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 4,dpi =300,limitsize = TRUE,bg = NULL,)

```


## Slingshot pseudotime analysis
```{r slingshot}
#omega sets the maximum allowable distance for connecting clusters (omega_scale=1.5 is standard)
lin <- getLineages(sce, reducedDim='PCA', clusterLabels='subset', omega=TRUE, omega_scale=0.9)
lin@int_metadata

#without starting cluster
sce <- slingshot(sce, reducedDim = 'PCA', clusterLabels = 'subset', start.clus=c('4. CD56bright'), omega=TRUE, omega_scale=0.9)

embedded <- embedCurves(sce, "UMAP", approx_points=FALSE, shrink=0)
embedded1 <- slingCurves(embedded)[[1]] # only 1 path.
embedded1 <- data.frame(embedded1$s[embedded1$ord,])

embedded2 <- slingCurves(embedded)[[2]] # only 1 path.
embedded2 <- data.frame(embedded2$s[embedded2$ord,])

embedded3 <- slingCurves(embedded)[[3]] # only 1 path.
embedded3<- data.frame(embedded3$s[embedded3$ord,])

embedded4<- slingCurves(embedded)[[4]] # only 1 path.
embedded4<- data.frame(embedded4$s[embedded4$ord,])


plotUMAP(sce, colour_by="sublabel")+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotUMAP(sce, colour_by="slingPseudotime_1") + geom_path(data=embedded1, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 13), axis.text=element_text(size=13))

plotUMAP(sce, colour_by="sublabel") + geom_path(data=embedded1, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotUMAP(sce, colour_by="slingPseudotime_2") + geom_path(data=embedded2, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 13), axis.text=element_text(size=13))
plotUMAP(sce, colour_by="sublabel") + geom_path(data=embedded2, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotUMAP(sce, colour_by="slingPseudotime_3") + geom_path(data=embedded3, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15))
plotUMAP(sce, colour_by="sublabel") + geom_path(data=embedded3, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")


plotUMAP(sce, colour_by="slingPseudotime_4") + geom_path(data=embedded4, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15))
plotUMAP(sce, colour_by="sublabel") + geom_path(data=embedded4, aes(x=Dim.1, y=Dim.2), size=1.2)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")


plotUMAP(sce, colour_by="color2", point_size=2) + geom_path(data=embedded1, aes(x=Dim.1, y=Dim.2), size=1.2)
plotUMAP(sce, colour_by="color2", point_size=2) + geom_path(data=embedded1, aes(x=Dim.1, y=Dim.2), size=1.2, point_alpha=1)

ggsave("lin1_pseudotime.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 4,dpi =300,limitsize = TRUE,bg = NULL,)

```

```{r}
saveRDS(sce, "sce_pseudotime.rds")
```


## Calculate which genes are driving the pseudotime
```{r}
set.seed(6)

# fit negative binomial GAM only for lineage 1 (to extract circulating NK cells, recreate slingshot)
counts(sce, withDimnames=FALSE) <- as(counts(sce, withDimnames=FALSE),"dgCMatrix")
sce <- sce[, sce$subset%in% c('1. CD56dim GZMK-', '2. CD56dimGZMK+', '4. CD56bright')]

#recalculate PCs, to do!!
set.seed(100) # See below.
sce <- runPCA(sce, subset_row=hvg.sce.var, name="PCA")

sce <- slingshot(sce, reducedDim = 'PCA', clusterLabels = 'subset', start.clus=c('4. CD56bright'), omega=TRUE, omega_scale=1.5)
sce <- fitGAM(sce, parallel=FALSE, genes=hvg.sce.var, nknots=6, verbose=TRUE)


# testing whether the average gene expression is significantly changing along pseudotime.
ATres <- associationTest(sce, lineages=TRUE)
head(ATres)

saveRDS(ATres, 'ATres.rds')

#select topgenes based on waldStatistic
topgenes_1 <- rownames(ATres[order(ATres$waldStat_1, decreasing=TRUE), ])[1:50]

#heatmap with binned pseudotime values, per lineage
yhatSmooth_1 <- predictSmooth(sce, gene = topgenes_1, nPoints = 100, tidy=FALSE)

pheatmap(t(scale(t(yhatSmooth_1[, 1:100]))),cluster_cols = FALSE,show_rownames = TRUE, show_colnames = FALSE,fontsize=8, border_color=NA)
plotSmoothers(sce, assays(sce)$counts, gene = "PRF1", alpha = 1, border = TRUE) + ggtitle("PRF1")
plotSmoothers(sce, assays(sce)$counts, gene = "GZMK", alpha = 1, border = TRUE) + ggtitle("GZMK")

plotExpression(sce, "FGFBP2", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('4. CD56bright'='#ed665d', '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotExpression(sce, "PRF1", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('4. CD56bright'='#ed665d',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotExpression(sce, "GZMB", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('4. CD56bright'='#ed665d', '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotExpression(sce, "CD44", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('4. CD56bright'='#ed665d', '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotExpression(sce, "TCF7", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('4. CD56bright'='#ed665d', '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")

plotExpression(sce, "GZMK", x = "slingPseudotime_1", colour_by = "sublabel", show_violin = FALSE,show_smooth = TRUE)+ theme(text = element_text(size = 15), axis.text=element_text(size=15)) + scale_color_manual(values = c('3. ltNK' = "#67bf5c", '5. CD34+' = "#ad8bc9", '4. CD56bright'='#ed665d', '6. ILC'='#a8786e',  '2. CD56dimGZMK+'='#ff9e4a', 'CD56dimGZMKmin_1' = "#A0CBE8", 'CD56dimGZMKmin_2' = "#D37295", 'CD56dimGZMKmin_3' = "#8CD17D", 'CD56dimGZMKmin_4'='#F1CE63', 'CD56dimGZMKmin_5'='#79706E', 'CD56dimGZMKmin_6'='#499894'),aesthetics = "fill")


ggsave("GZMK.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 4,dpi =300,limitsize = TRUE,bg = NULL,)
```

## RNA velocity

```{r}
#(read velocity is dependent on linux)
loom <- ReadVelocity(file = "/exports/klim-hpc/Janine/Janine_GEX_count_7k/velocyto/Janine_GEX_count_7k.loom")
seurat_velo <- as.Seurat(x = loom)

#sce file without CD56- and NKprolif
sce <- readRDS("I:/Laboratoria/immlab/Janine/singelcellRNAseq/submitted/Nature comm/rebuttal/Rscripts/sce_pseudotime.rds")
counts(sce, withDimnames=FALSE) <- as(counts(sce, withDimnames=FALSE), "dgCMatrix")
logcounts(sce, withDimnames=FALSE) <- as(logcounts(sce, withDimnames=FALSE), "dgCMatrix")

df <- as.data.frame(sce@colData)
rownames(df) <- df$barcode
seurat <- CreateSeuratObject(sce@assays@data@listData[["counts"]] , meta.data = df)
seurat@meta.data$subset <- df$subset

# Rename cells in a Seurat object
seurat_velo <- readRDS("P:/kjc/Immlab/janineM/10xGenomics/Janine_GEX_count_7k/sceobjects/BMKJC/seurat_velo.rds")
seurat_velo <- RenameCells(seurat_velo, new.names= gsub("Janine_GEX_count_7k:", "", Cells(seurat_velo)))
seurat_velo <- RenameCells(seurat_velo, new.names= gsub("x", "-1", Cells(seurat_velo)))

#subset seurat_velo 
seurat_velo <- subset(x = seurat_velo, cells=Cells(seurat))

#set reductions seurat to seurat_velo
seurat_velo@meta.data <- seurat@meta.data
seurat_velo@reductions <- seurat@reductions

#check ratio spliced/unspliced
spliced <- colSums(seurat_velo@assays$spliced)
unspliced <- colSums(seurat_velo@assays$unspliced)
ambiguous <- colSums(seurat_velo@assays$ambiguous)
spliced <- as.data.frame(spliced)
unspliced <- as.data.frame(unspliced)
ambiguous <- as.data.frame(ambiguous)
ratio <- spliced/unspliced
perc_unspliced <- unspliced/(spliced+unspliced)*100
seurat_velo <- AddMetaData(object = seurat_velo, metadata = ratio, col.name = "ratio")
seurat_velo <- AddMetaData(object = seurat_velo, metadata = perc_unspliced, col.name = "perc_unspliced")

VlnPlot(seurat_velo, features='ratio', group.by='subset')
VlnPlot(seurat_velo, features='perc_unspliced', group.by='subset')

#rewrite seurat to sce object
sce_velo_spliced <- as.SingleCellExperiment(seurat_velo, assay='spliced')
sce_velo_unspliced <- as.SingleCellExperiment(seurat_velo, assay='unspliced')
sce_velo <- SingleCellExperiment(list(spliced=counts(sce_velo_spliced), unspliced = counts(sce_velo_unspliced), exon=counts(sce), logcounts=logcounts(sce)))

colData(sce_velo) <- colData(sce)
reducedDim(sce_velo, 'UMAP', withDimnames=FALSE) = reducedDim(sce, 'UMAP', withDimnames=FALSE)
reducedDim(sce_velo, 'PCA', withDimnames=FALSE) = reducedDim(sce, 'PCA', withDimnames=FALSE)
#reducedDim(sce_velo, "UMAP") <- reducedDim(sce, "UMAP")


#or run scvelo, x= count matrix sce (unnormalized)
velo.out<- scvelo(sce_velo, assay.spliced='spliced', assay.unspliced='unspliced', assay.X='exon', use.dimred='PCA', mode='stochastic')
velo.out$subset <- sce_velo$sublabel

sce_velo$velocity_pseudotime <- velo.out$velocity_pseudotime
embedded <- embedVelocity(reducedDim(sce_velo, 'UMAP'), velo.out)
grid.df <- gridVectors(reducedDim(sce_velo, 'UMAP'), embedded, resolution=30)

#diagnostic plots
plotVelocity(velo.out, genes=c('FCGR3A'), use.dimred='UMAP', assay.splicedM='spliced', assay.unsplicedM='unspliced', which.plots=c('phase', 'velocity', 'expression'), color_by='subset')
plotColData(sce_velo, x='subset', y='velocity_pseudotime', colour_by = 'subset')

reducedDim(velo.out, "UMAP") <- reducedDim(sce_velo, "UMAP")
plotReducedDim(sce_velo,dimred="UMAP", colour_by ="velocity_pseudotime", point_alpha=0.3)+geom_segment(data=grid.df, mapping=aes(x=start.1, y=start.2, xend=end.1, yend=end.2), arrow=arrow(length=unit(0.05, 'inches'), type="closed"))+ theme(text = element_text(size = 13), axis.text=element_text(size=13))

ggsave("rna_velocity.tiff",plot = last_plot(),device = NULL,path = NULL,scale = 1,width = 6,height = 4,dpi =300,limitsize = TRUE,bg = NULL,)

```

